docs/internals.md

```diff
@@ -0,0 +1,93 @@
+(e.g., unnecessary matching on subsamples of two different datasets).
+
+## Vfunc
+
+When a `Vset` is initialized, the items in the `modules` arg are wrapped in
+`Vfunc` objects, which are like named functions that can optionally support a
+`fit` or `transform` method. If a `Vset` is initialized with `is_async=True`,
+then modules are wrapped in `AsyncModule` objects, which compute asynchronously
+using `ray`. If a `Vset` is initialized with `lazy=True`, `Vset` output
+dictionary values are wrapped in `VfuncPromise` objects, which are lazily
+evaluated. At the moment, `VfuncPromise` objects are only resolved if passed to
+a `Vset` with `lazy=False` downstream or if called manually by the user.
+
+## Subkey
+
+By default, `vflow.subkey.Subkey` objects are non-matching, meaning that `vflow`
+won't bother to look for a matching `Subkey` when combining data or deciding
+which `Vfunc` to apply to which inputs. As described above, `Vset` inputs are
+combined from left to right two-at-a-time, and if every `Subkey` in the keys of
+both input dictionaries is non-matching, then the result is the full Cartesian
+product of the two dictionaries. (Note, however, that later inputs where
+matching is required may invalidate some entries in that Cartesian product.)
+There is one main way for a `Subkey` to be matching: it was created by a `Vset`
+that was initialized with `output_matching=True`.
+
+When `vflow` combines two dictionaries it previously created, if one of the
+entries in a given `Subkey` tuple key is matching then `vflow` tries to find a
+match in the tuple keys of the other dictionary. If the other dictionary's
+`Subkey`s are non-matching, or if there is a matching `Subkey` with the same
+origin and same value, then the keys are combined and their values are either
+combined (when combining inputs) or a `Vfunc` is applied to matching input (when
+using previously combined inputs). For example, say we have a data cleaning
+`Vset` named "cleaning" that was initialized with `output_matching=True` and has
+a `Vfunc` named "median impute" and another named "mean impute", and say this
+`Vset` was applied to a training set and later to a test set. When the `Vset` is
+applied to the training data, the output dictionary of cleaned datasets has some
+tuple keys containing a `Subkey` with origin "cleaning" and value "median
+impute", and that `Subkey` is carried through when the cleaned data is passed to
+a modeling `Vset`. When we want to produce test predictions from the modeling
+`Vset`, `vflow` decides which trained models to apply to which version of the
+cleaned test data based on the `Subkey`s created by the data cleaning `Vset`. We
+have a positive match if the tuple keys each have a `Subkey` with origin
+"cleaning" and value "median impute". On the other hand, if one of the `Subkey`s
+instead had value "mean impute", then the two keys mismatch and we proceed to
+check the next dictionary key without applying the trained model to the test
+data.
```

Comment:
Might be easier to follow by using subheadings in this paragraph. Could be something like:

#### When `vflow` combines two dictionaries it previously created
...

#### When the `Vset` is applied to the training data
...

#### When the modeling `Vset` produces test predictions
...